planner/core/logical_plan_builder.go

```diff
@@ -3290,12 +3290,21 @@ func (b *PlanBuilder) buildSelect(ctx context.Context, sel *ast.SelectStmt) (p L
 		projExprs                     []expression.Expression
 	)
 
-	// For sub-queries, the FROM clause may have already been built in outer query when resolving correlated aggregates.
-	// If the ResultSetNode inside FROM clause has nothing to do with correlated aggregates, we can simply get the
-	// existing ResultSetNode from the cache.
-	p, err = b.buildTableRefsWithCache(ctx, sel.From)
-	if err != nil {
-		return nil, err
+	// set for update read to true before building result set node
+	if isForUpdateReadSelectLock(sel.LockInfo) {
+		b.isForUpdateRead = true
+	}
+
+	if sel.From != nil {
+		// For sub-queries, the FROM clause may have already been built in outer query when resolving correlated aggregates.
+		// If the ResultSetNode inside FROM clause has nothing to do with correlated aggregates, we can simply get the
+		// existing ResultSetNode from the cache.
+		p, err = b.buildTableRefsWithCache(ctx, sel.From)
+		if err != nil {
+			return nil, err
+		}
+	} else {
```

Comment:
why add this `if else`~?

it seems `buildTableRefsWithCache` will return tableDual when `sel.From == nil` 

https://github.com/pingcap/tidb/blob/b45fa8b891b80376b37ab39759730310a182ae5a/planner/core/logical_plan_builder.go#L309